File: exam_in_class/1/one_more_rail_problem.py
#
# author: jamie
# email: jinjiedeng.jjd@gmail.com
#
from collections import deque

def Solve (A, B, S):
    while len(A):
        S.append(A[0])
        A.popleft()
        while len(S) and (S[0] == B[0] or S[-1] == B[0]):
            if S[0] == B[0]:
                S.popleft()
                B.popleft()
            elif len(S) and S[-1] == B[0]:
                S.pop()
                B.popleft()
    print("NO" if len(B) else "YES")

if __name__ == "__main__":
    m = int(input())
    for _ in range(m):
        n = int(input())
        A = deque(map(int, input().split()))
        B = deque(map(int, input().split()))
        S = deque()
        Solve(A, B, S)

File: exam_in_class/1/one_more_rail_problem.cc
/*
* author: jamie
* email: jinjiedeng.jjd@gmail.com
*/
#include <iostream>
#include <deque>

int main ()
{
  uint64_t m;
  std::cin >> m;
  while (m--)
  {
    uint64_t n;
    std::cin >> n;
    std::deque<uint64_t> A, B, S;
    for (uint64_t i {}, a; i != n; ++i) { std::cin >> a; A.push_back(a); }
    for (uint64_t i {}, b; i != n; ++i) { std::cin >> b; B.push_back(b); }
    while (!A.empty())
    {
      S.push_back(A.front()); A.pop_front();
      while (!S.empty() && (S.front() == B.front() || S.back() == B.front()))
      {
        if (S.front() == B.front())               { S.pop_front(); B.pop_front(); }
        if (!S.empty() && S.back() == B.front())  { S.pop_back();  B.pop_front(); }
      }
    }
    std::cout << ((B.empty()) ? "YES" : "NO") << "\n";
  }
  return 0;
}